File: app/src/main/java/com/ravisharma/playbackmusic/adapters/NowPlayingAdapter.java
package com.ravisharma.playbackmusic.adapters;

import android.content.Context;
import android.net.Uri;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.recyclerview.widget.RecyclerView;

import com.airbnb.lottie.LottieAnimationView;
import com.bumptech.glide.Glide;
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions;
import com.bumptech.glide.request.RequestOptions;
import com.ravisharma.playbackmusic.MainActivity;
import com.ravisharma.playbackmusic.model.Song;
import com.ravisharma.playbackmusic.R;

import java.util.ArrayList;
import java.util.concurrent.TimeUnit;


/**
 * Created by Ravi Sharma on 07-Jan-18.
 */

public class NowPlayingAdapter extends RecyclerView.Adapter<NowPlayingAdapter.ViewHolder> {

    private ArrayList<Song> songs;
    Context c;

    //declare interface
    private OnItemClicked onClick;

    public void setList(ArrayList<Song> songList) {
        this.songs = songList;
        notifyDataSetChanged();
    }

    //make interface like this
    public interface OnItemClicked {
        void onItemClick(int position);

        void onOptionsClick(int position);
    }

    public NowPlayingAdapter(ArrayList<Song> theSongs, Context c) {
        songs = theSongs;
        this.c = c;
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        TextView title;
        TextView artist;
        TextView durView;
        ImageView songart;
        ImageView imgOptions;
        RelativeLayout songbox;
        LottieAnimationView animationView;

        public ViewHolder(View itemView) {
            super(itemView);
            title = itemView.findViewById(R.id.song_title);
            artist = itemView.findViewById(R.id.song_artist);
            durView = itemView.findViewById(R.id.duration);
            songbox = itemView.findViewById(R.id.song_box);
            songart = itemView.findViewById(R.id.song_art);
            imgOptions = itemView.findViewById(R.id.imgOptions);
            animationView = itemView.findViewById(R.id.animation_view);
        }
    }

    @Override
    public NowPlayingAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.adap_now_playing, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(NowPlayingAdapter.ViewHolder holder, final int position) {
        Song currSong = songs.get(position);
        TextView title = holder.title;
        TextView artist = holder.artist;
        TextView duration = holder.durView;
        ImageView art = holder.songart;

        title.setText(currSong.getTitle());
        artist.setText(currSong.getArtist());
        duration.setText((String.format("%d:%02d",
                TimeUnit.MILLISECONDS.toMinutes(currSong.getDuration()),
                TimeUnit.MILLISECONDS.toSeconds(currSong.getDuration()) -
                        TimeUnit.MINUTES.toSeconds(TimeUnit.MILLISECONDS.toMinutes(currSong.getDuration())))));

        /*Song art code here*/
        RequestOptions requestOptions = new RequestOptions();
        requestOptions.placeholder(R.drawable.logo);
        requestOptions.error(R.drawable.logo);

        Glide.with(c)
                .setDefaultRequestOptions(requestOptions)
                .load(Uri.parse(currSong.getArt()))
                .into(art);

        holder.songbox.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onClick.onItemClick(position);
            }
        });

        holder.imgOptions.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onClick.onOptionsClick(position);
            }
        });
    }

    @Override
    public int getItemCount() {
        return songs.size();
    }

    public void setOnClick(OnItemClicked onClick) {
        this.onClick = onClick;
    }
}

File: app/src/main/java/com/ravisharma/playbackmusic/adapters/NowPlayingAdapter.kt
package com.ravisharma.playbackmusic.adapters

import android.net.Uri
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.ViewGroup
import androidx.annotation.Nullable
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import coil.load
import coil.transform.RoundedCornersTransformation
import com.ravisharma.playbackmusic.R
import com.ravisharma.playbackmusic.databinding.AdapNowPlayingBinding
import com.ravisharma.playbackmusic.data.olddb.model.Song
import com.ravisharma.playbackmusic.utils.StartDragListener
import java.util.concurrent.TimeUnit

class NowPlayingAdapter(private var dragListener: StartDragListener) : RecyclerView.Adapter<NowPlayingAdapter.ViewHolder>() {

    private var songs: ArrayList<Song> = ArrayList()

    //declare interface
    private lateinit var onClick: OnItemClicked

    fun setList(songList: ArrayList<Song>) {
        val diffCallback = DiffCallback(songs, songList)
        val diffResult = DiffUtil.calculateDiff(diffCallback)
        songs.clear()
        songs.addAll(songList)
        diffResult.dispatchUpdatesTo(this)
    }

    //make interface like this
    interface OnItemClicked {
        fun onItemClick(position: Int)
        fun onOptionsClick(position: Int)
        fun onItemRemove(position: Int)
    }

    inner class ViewHolder(val binding: AdapNowPlayingBinding) : RecyclerView.ViewHolder(binding.root)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding = AdapNowPlayingBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val currSong = songs[holder.adapterPosition]
        holder.binding.apply {
            songTitle.text = currSong.title
            songArtist.text = currSong.artist
            duration.text = String.format("%d:%02d",
                    TimeUnit.MILLISECONDS.toMinutes(currSong.duration),
                    TimeUnit.MILLISECONDS.toSeconds(currSong.duration) -
                            TimeUnit.MINUTES.toSeconds(TimeUnit.MILLISECONDS.toMinutes(currSong.duration)))

            songArt.load(Uri.parse(currSong.art)) {
                placeholder(R.drawable.logo)
                error(R.drawable.logo)
                transformations(RoundedCornersTransformation(20f))
            }

            songBox.setOnClickListener { onClick.onItemClick(holder.adapterPosition) }
            songBox.setOnLongClickListener {
                onClick.onOptionsClick(holder.adapterPosition)
                true
            }

//            imgRemove.setOnClickListener {
//                onClick.onItemRemove(holder.adapterPosition)
//            }

            ivOrder.setOnTouchListener { v, event ->
                if (event.action == MotionEvent.ACTION_DOWN) {
                    dragListener.requestDrag(holder)
                }
                true
            }
        }
    }

    override fun getItemCount() = songs.size

    fun setOnClick(onClick: OnItemClicked) {
        this.onClick = onClick
    }

    inner class DiffCallback(private val oldList: ArrayList<Song>, private val newList: ArrayList<Song>) : DiffUtil.Callback() {
        override fun getOldListSize(): Int = oldList.size

        override fun getNewListSize(): Int = newList.size

        override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
            return oldList[oldItemPosition].id == newList[newItemPosition].id
        }

        override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
            return oldList[oldItemPosition].id == newList[newItemPosition].id
        }

        @Nullable
        override fun getChangePayload(oldItemPosition: Int, newItemPosition: Int): Any? {
            return super.getChangePayload(oldItemPosition, newItemPosition)
        }
    }
}